Secondary structure of the RNA molecule is actively involved in different genome regulation mechanisms, moreover, it contains a lot of important information for sequences phylogenetic and taxonomic analysis. Therefore, RNA secondary structure prediction problem is quite critical in computational genomics, unfortunately, it still remains open due to the complexity and variability of secondary structure formation laws along with a high degree of noisiness in biological data. Another challenging part here is the processing of pseudoknotted structures that are known to be widely represented in organisms genomes, including functionally important RNA regions, nevertheless, the overlapping nature of pseudoknots makes handling them an NP-complete task~\cite{lyngso2000rna}. 

The best results in the secondary structure prediction field belong to the laboratory methods, such as X-ray structural analysis~\cite{westhof2015twenty} and nuclear magnetic resonance~\cite{furtig2003nmr}, however, the high cost and complexity of such experiments lead to the rapid development of computational methods all the diversity of which can be divided into two groups. The first one contains comparative methods that analyze several homologous sequences employing evolutionary approaches~\cite{hofacker1999automatic,tahi2002automatic}. These methods are known to be quite reliable but they demand a lot of data and complicated manual analysis. The second group aggregates different single sequence methods that process one sequence at a time applying some mathematical or physical model along with its optimization scheme. The most popular approach here is the minimum free energy (MFE) principle based on secondary structure thermodynamic stability, and the task of searching the MFE folding can be solved by dynamic programming~\cite{bellaousov2013rnastructure,rivas1999dynamic}, heuristic algorithms~\cite{ren2005hotknots,ruan2004ilm} and other optimization techniques~\cite{reeder2007pknotsrg,jabbari2018knotty}. Moreover, there are methods that do not demand any physically measurable parameters and model secondary structure by maximizing some function of expected accuracy (MEA)~\cite{sato2011ipknot,sato2009centroidfold} or building stochastic context-free grammar~\cite{knudsen1999rna,dowell2004evaluation,rivas1999dynamic}. RNA folding mechanisms are known to be non-trivial for understanding and formalization, therefore, machine learning techniques are widely used for secondary structure prediction either integrated into some strict algorithm pipeline~\cite{akiyama2018max,do2006contrafold} or as the main method itself~\cite{singh2019rna,apolloni2003rna}. 

In this work, we introduce a new approach for RNA secondary structure prediction employing the combination of ordinary formal grammars and artificial neural networks. The main ideas were outlined in~\cite{grigorevcomposition,lunina2019secondary} and this research is conducted to the further development of this approach in the context of secondary structure prediction problem. The classical way of formal languages application here is to use complicated stochastic grammars for modeling the whole structure~\cite{knudsen1999rna,dowell2004evaluation,rivas2000language} and this approach is known to be quite successful, however, building such grammar requires a lot of theoretical and practical difficulties. To overcome these issues we propose to split the secondary structure into elementary parts (stems), describe them by very simple grammar, and use neural networks to synthesize the final result. So, on the one hand, using neural networks allows to skip full formalization of RNA secondary structure and on the other hand, the grammar provides some sort of basis for neural network training. 